File: leetcode/lc-us/cpp/543.cpp
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}

    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}

    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
    int ans = 1;

    int depth(TreeNode *node) {
        if (!node) return 0;
        int l = depth(node->left);
        int r = depth(node->right);
        if (l + r + 1 > ans) ans = l + r + 1;
        return l > r ? l + 1 : r + 1;
    }

public:
    int diameterOfBinaryTree(TreeNode *root) {
        depth(root);
        return ans - 1;
    }
};

File: leetcode/lc-us/py/543.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def __init__(self):
        self.res = 0

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def dfs(node: Optional[TreeNode]) -> int:
            if not node:
                return 0
            left, right = dfs(node.left), dfs(node.right)
            self.res = max(self.res, left + right)
            return max(left, right) + 1

        dfs(root)
        return self.res
